File: apps/carbon/lib/api/intelligence.ts
/**
 * lib/api/intelligence.ts — client des endpoints `/intelligence/*` + fraîcheur
 * (PR-04, noyau Evidence Kernel PR-03).
 *
 * Réutilise `API_BASE_URL` + le porteur de token en mémoire de `lib/api.ts`
 * (JWT posé par use-auth). Types en snake_case, miroir de
 * `apps/api/models/intelligence.py`.
 */

import { API_BASE_URL, getAuthToken } from "@/lib/api";

// ---------------------------------------------------------------------------
// Types (miroir models/intelligence.py)
// ---------------------------------------------------------------------------

export type SourceType = "api" | "file" | "webpage" | "manual" | "licensed_feed";
export type ReleaseStatus =
  | "detected"
  | "quarantined"
  | "validated"
  | "published"
  | "superseded"
  | "blocked_license";
export type BackendDataStatus = "verified" | "estimated" | "manual" | "inferred";

export interface IntelligenceSource {
  id: number;
  company_id: number | null;
  code: string;
  publisher: string;
  title: string;
  source_type: SourceType;
  adapter_kind: string | null;
  base_uri: string | null;
  license_code: string | null;
  automated_access_allowed: boolean;
  storage_allowed: boolean;
  commercial_use_allowed: boolean;
  redistribution_allowed: boolean;
  derived_use_allowed: boolean;
  display_allowed: boolean;
  attribution_text: string | null;
  terms_uri: string | null;
  active: boolean;
  created_by: number | null;
  created_at: string;
  updated_at: string;
}

export interface SourceListResponse {
  items: IntelligenceSource[];
  total: number;
  limit: number;
  offset: number;
}

export interface Release {
  id: number;
  source_id: number;
  company_id: number | null;
  release_key: string;
  published_at: string | null;
  retrieved_at: string;
  valid_from: string | null;
  valid_to: string | null;
  checksum_sha256: string;
  blob_key: string | null;
  mime_type: string | null;
  schema_version: string | null;
  status: ReleaseStatus;
  supersedes_id: number | null;
  metadata: Record<string, unknown>;
  created_by: number | null;
  created_at: string;
}

export interface ReleaseListResponse {
  items: Release[];
  total: number;
  limit: number;
  offset: number;
}

export interface Observation {
  id: number;
  company_id: number | null;
  subject_type: string;
  subject_key: string;
  metric_code: string;
  numeric_value: number | null;
  text_value: string | null;
  boolean_value: boolean | null;
  unit: string | null;
  geography_code: string | null;
  stage_code: string | null;
  observed_at: string | null;
  valid_from: string | null;
  valid_to: string | null;
  source_release_id: number;
  evidence_artifact_id: number | null;
  data_status: BackendDataStatus;
  confidence: number | null;
  methodology_version: string | null;
  supersedes_id: number | null;
  created_at: string;
}

export interface ObservationListResponse {
  items: Observation[];
  total: number;
  limit: number;
  offset: number;
}

export interface SourceFreshness {
  source_id: number;
  company_id: number | null;
  code: string;
  publisher: string;
  title: string;
  source_type: SourceType;
  active: boolean;
  last_release_id: number | null;
  last_release_key: string | null;
  last_release_status: ReleaseStatus | null;
  last_release_at: string | null;
  published_release_count: number;
  total_release_count: number;
  has_release: boolean;
  age_days: number | null;
  is_stale: boolean;
  license_ok: boolean;
  allow_display: boolean;
  allow_derived_use: boolean;
  license_reasons: string[];
  license_warnings: string[];
}

export interface SourceFreshnessListResponse {
  items: SourceFreshness[];
  total: number;
  limit: number;
  offset: number;
}

export interface IntelligenceHealthSource {
  code: string;
  last_release_at: string | null;
  age_days: number | null;
  last_release_status: ReleaseStatus | null;
  is_stale: boolean;
  license_ok: boolean;
}

export interface IntelligenceHealth {
  status: "ok" | "degraded" | "empty";
  checked_at: string;
  source_count: number;
  stale_count: number;
  license_anomaly_count: number;
  sources: IntelligenceHealthSource[];
  db: "ok" | "not_configured" | "down";
}

// ---------------------------------------------------------------------------
// Fetch helpers — token JWT en mémoire (posé par use-auth)
// ---------------------------------------------------------------------------

function authHeaders(): Record<string, string> {
  const token = getAuthToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function get<T>(path: string, signal?: AbortSignal): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "GET",
    headers: { Accept: "application/json", ...authHeaders() },
    credentials: "include",
    signal,
  });
  if (!res.ok) throw new Error(`API ${res.status} on ${path}`);
  return (await res.json()) as T;
}

async function post<T>(path: string, body?: unknown, signal?: AbortSignal): Promise<T> {
  const headers: Record<string, string> = { Accept: "application/json", ...authHeaders() };
  if (body !== undefined) headers["Content-Type"] = "application/json";
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
    headers,
    body: body !== undefined ? JSON.stringify(body) : undefined,
    credentials: "include",
    signal,
  });
  if (!res.ok) throw new Error(`API ${res.status} on ${path}`);
  return (await res.json()) as T;
}

// ---------------------------------------------------------------------------
// Endpoints
// ---------------------------------------------------------------------------

export function fetchSources(
  opts: { limit?: number; offset?: number; activeOnly?: boolean } = {},
  signal?: AbortSignal,
): Promise<SourceListResponse> {
  const p = new URLSearchParams();
  if (opts.limit != null) p.set("limit", String(opts.limit));
  if (opts.offset != null) p.set("offset", String(opts.offset));
  if (opts.activeOnly) p.set("active_only", "true");
  const qs = p.toString() ? `?${p.toString()}` : "";
  return get<SourceListResponse>(`/intelligence/sources${qs}`, signal);
}

export function fetchSource(id: number, signal?: AbortSignal): Promise<IntelligenceSource> {
  return get<IntelligenceSource>(`/intelligence/sources/${id}`, signal);
}

export function fetchSourceReleases(
  id: number,
  opts: { limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<ReleaseListResponse> {
  const p = new URLSearchParams();
  if (opts.limit != null) p.set("limit", String(opts.limit));
  if (opts.offset != null) p.set("offset", String(opts.offset));
  const qs = p.toString() ? `?${p.toString()}` : "";
  return get<ReleaseListResponse>(`/intelligence/sources/${id}/releases${qs}`, signal);
}

export function fetchObservations(
  opts: { subjectType?: string; subjectKey?: string; metricCode?: string; limit?: number; offset?: number } = {},
  signal?: AbortSignal,
): Promise<ObservationListResponse> {
  const p = new URLSearchParams();
  if (opts.subjectType) p.set("subject_type", opts.subjectType);
  if (opts.subjectKey) p.set("subject_key", opts.subjectKey);
  if (opts.metricCode) p.set("metric_code", opts.metricCode);
  if (opts.limit != null) p.set("limit", String(opts.limit));
  if (opts.offset != null) p.set("offset", String(opts.offset));
  const qs = p.toString() ? `?${p.toString()}` : "";
  return get<ObservationListResponse>(`/intelligence/observations${qs}`, signal);
}

export function fetchSourceFreshness(id: number, signal?: AbortSignal): Promise<SourceFreshness> {
  return get<SourceFreshness>(`/intelligence/sources/${id}/freshness`, signal);
}

/** État public de fraîcheur (sources globales) — ne nécessite pas d'auth. */
export function fetchIntelligenceHealth(signal?: AbortSignal): Promise<IntelligenceHealth> {
  return get<IntelligenceHealth>(`/health/intelligence`, signal);
}

// Transitions de release (require_admin côté API).
export function validateRelease(id: number, passed = true, signal?: AbortSignal): Promise<Release> {
  return post<Release>(`/intelligence/releases/${id}/validate`, { passed }, signal);
}

export function publishRelease(id: number, signal?: AbortSignal): Promise<Release> {
  return post<Release>(`/intelligence/releases/${id}/publish`, undefined, signal);
}

export function supersedeRelease(id: number, signal?: AbortSignal): Promise<Release> {
  return post<Release>(`/intelligence/releases/${id}/supersede`, undefined, signal);
}

File: apps/api/models/intelligence.py
"""
models/intelligence.py — Evidence Kernel (PR-03) : sources, releases,
artefacts, ingestions, observations, liens de preuve.

Un modèle par table + un modèle de requête par opération d'écriture exposée.
Champs en snake_case, alignés sur les colonnes SQL de la migration 028 (pas
de camelCase ici — ce noyau n'a encore aucun consommateur frontend, cf.
routers/facts.py qui suit la même convention pour la donnée API interne).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

SourceType = Literal["api", "file", "webpage", "manual", "licensed_feed"]
ReleaseStatus = Literal[
    "detected", "quarantined", "validated", "published", "superseded", "blocked_license"
]
IngestionStatus = Literal[
    "pending", "running", "quarantined", "validated", "published", "failed", "blocked_license"
]
ArtifactSensitivity = Literal["public", "internal", "confidential", "restricted"]
DataStatus = Literal["verified", "estimated", "manual", "inferred"]
ClaimRelationType = Literal["supports", "contradicts", "contextualizes", "derived_from"]


# ---------------------------------------------------------------------------
# source_registry
# ---------------------------------------------------------------------------

class SourceCreate(BaseModel):
    code: str
    publisher: str
    title: str
    source_type: SourceType
    adapter_kind: str | None = None
    base_uri: str | None = None
    license_code: str | None = None
    automated_access_allowed: bool = False
    storage_allowed: bool = False
    commercial_use_allowed: bool = False
    redistribution_allowed: bool = False
    derived_use_allowed: bool = False
    display_allowed: bool = False
    attribution_text: str | None = None
    terms_uri: str | None = None


class SourceUpdate(BaseModel):
    """PATCH — tous les champs optionnels, seuls les champs fournis sont modifiés."""
    publisher: str | None = None
    title: str | None = None
    adapter_kind: str | None = None
    base_uri: str | None = None
    license_code: str | None = None
    automated_access_allowed: bool | None = None
    storage_allowed: bool | None = None
    commercial_use_allowed: bool | None = None
    redistribution_allowed: bool | None = None
    derived_use_allowed: bool | None = None
    display_allowed: bool | None = None
    attribution_text: str | None = None
    terms_uri: str | None = None
    active: bool | None = None


class SourceResponse(BaseModel):
    id: int
    company_id: int | None
    code: str
    publisher: str
    title: str
    source_type: SourceType
    adapter_kind: str | None
    base_uri: str | None
    license_code: str | None
    automated_access_allowed: bool
    storage_allowed: bool
    commercial_use_allowed: bool
    redistribution_allowed: bool
    derived_use_allowed: bool
    display_allowed: bool
    attribution_text: str | None
    terms_uri: str | None
    active: bool
    created_by: int | None
    created_at: datetime
    updated_at: datetime


class SourceListResponse(BaseModel):
    items: list[SourceResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# source_releases
# ---------------------------------------------------------------------------

class ReleaseCreate(BaseModel):
    release_key: str
    checksum_sha256: str = Field(min_length=64, max_length=64)
    published_at: datetime | None = None
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    blob_key: str | None = None
    mime_type: str | None = None
    schema_version: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    # Fixé à la création, jamais modifié ensuite (evidence_kernel_guard le
    # gèle avec le reste de la ligne une fois published) — la release qui
    # CORRIGE une autre déclare le lien dès sa détection, pas après coup.
    supersedes_id: int | None = None


class ReleaseResponse(BaseModel):
    id: int
    source_id: int
    company_id: int | None
    release_key: str
    published_at: datetime | None
    retrieved_at: datetime
    valid_from: datetime | None
    valid_to: datetime | None
    checksum_sha256: str
    blob_key: str | None
    mime_type: str | None
    schema_version: str | None
    status: ReleaseStatus
    supersedes_id: int | None
    metadata: dict[str, Any]
    created_by: int | None
    created_at: datetime


class ReleaseListResponse(BaseModel):
    items: list[ReleaseResponse]
    total: int
    limit: int
    offset: int


class ReleaseValidateRequest(BaseModel):
    """Corps de POST /releases/{id}/validate — `passed=false` met en quarantaine."""
    passed: bool = True


# ---------------------------------------------------------------------------
# evidence_artifacts
# ---------------------------------------------------------------------------

class ArtifactResponse(BaseModel):
    id: int
    company_id: int | None
    source_release_id: int | None
    blob_key: str
    sha256: str
    filename: str
    mime_type: str
    size_bytes: int | None
    page_reference: str | None
    table_reference: str | None
    cell_reference: str | None
    excerpt: str | None
    sensitivity: ArtifactSensitivity
    created_by: int | None
    created_at: datetime


# ---------------------------------------------------------------------------
# ingestion_runs
# ---------------------------------------------------------------------------

class IngestionRunResponse(BaseModel):
    id: int
    company_id: int | None
    source_id: int
    source_release_id: int | None
    adapter_kind: str | None
    idempotency_key: str
    status: IngestionStatus
    detected_count: int
    accepted_count: int
    rejected_count: int
    warning_count: int
    error_summary: str | None
    started_at: datetime
    completed_at: datetime | None
    created_by: int | None
    metadata: dict[str, Any]


class IngestionRunListResponse(BaseModel):
    items: list[IngestionRunResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# observations
# ---------------------------------------------------------------------------

class ObservationCreate(BaseModel):
    subject_type: str
    subject_key: str
    metric_code: str
    numeric_value: float | None = None
    text_value: str | None = None
    boolean_value: bool | None = None
    unit: str | None = None
    geography_code: str | None = None
    stage_code: str | None = None
    observed_at: datetime | None = None
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    source_release_id: int
    evidence_artifact_id: int | None = None
    data_status: DataStatus
    confidence: float | None = Field(default=None, ge=0, le=1)
    methodology_version: str | None = None
    supersedes_id: int | None = None


class ObservationResponse(BaseModel):
    id: int
    company_id: int | None
    subject_type: str
    subject_key: str
    metric_code: str
    numeric_value: float | None
    text_value: str | None
    boolean_value: bool | None
    unit: str | None
    geography_code: str | None
    stage_code: str | None
    observed_at: datetime | None
    valid_from: datetime | None
    valid_to: datetime | None
    source_release_id: int
    evidence_artifact_id: int | None
    data_status: DataStatus
    confidence: float | None
    methodology_version: str | None
    supersedes_id: int | None
    created_at: datetime


class ObservationListResponse(BaseModel):
    items: list[ObservationResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# claim_evidence_links (schéma PR-03 ; service livré par PR-05A —
# claim_link_service.py, cf. WAVE_2_INTERFACE_CONTRACTS.md §1)
# ---------------------------------------------------------------------------

class ClaimEvidenceLinkCreate(BaseModel):
    claim_type: str = Field(min_length=1, max_length=100)
    claim_key: str = Field(min_length=1, max_length=255)
    evidence_artifact_id: int
    relation_type: ClaimRelationType


class ClaimEvidenceLinkResponse(BaseModel):
    id: int
    company_id: int | None
    claim_type: str
    claim_key: str
    evidence_artifact_id: int
    relation_type: ClaimRelationType
    created_by: int | None
    created_at: datetime


class ClaimEvidenceLinkListResponse(BaseModel):
    items: list[ClaimEvidenceLinkResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# license_policy — décision structurée (jamais un booléen nu)
# ---------------------------------------------------------------------------

class LicenseDecision(BaseModel):
    allow_ingest: bool
    allow_store: bool
    allow_display: bool
    allow_derived_use: bool
    reasons: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Fraîcheur des sources (PR-04) — vue source_freshness + licence évaluée
# ---------------------------------------------------------------------------

class SourceFreshness(BaseModel):
    source_id: int
    company_id: int | None
    code: str
    publisher: str
    title: str
    source_type: SourceType
    active: bool
    last_release_id: int | None
    last_release_key: str | None
    last_release_status: ReleaseStatus | None
    last_release_at: datetime | None
    published_release_count: int
    total_release_count: int
    has_release: bool
    age_days: int | None
    is_stale: bool
    # Licence évaluée déterministe (license_policy) — jamais un booléen nu.
    license_ok: bool
    allow_display: bool
    allow_derived_use: bool
    license_reasons: list[str] = Field(default_factory=list)
    license_warnings: list[str] = Field(default_factory=list)


class SourceFreshnessListResponse(BaseModel):
    items: list[SourceFreshness]
    total: int
    limit: int
    offset: int


class IntelligenceHealthSource(BaseModel):
    """Ligne minimale et publique (aucun secret) — sources GLOBALES uniquement."""
    code: str
    last_release_at: datetime | None
    age_days: int | None
    last_release_status: ReleaseStatus | None
    is_stale: bool
    license_ok: bool


class IntelligenceHealthResponse(BaseModel):
    status: Literal["ok", "degraded", "empty"]
    checked_at: datetime
    source_count: int
    stale_count: int
    license_anomaly_count: int
    sources: list[IntelligenceHealthSource]
    db: Literal["ok", "not_configured", "down"] = "ok"
